Makefile.rules

```diff
@@ -35,12 +46,41 @@ ifeq ($(NO_CXNG),)
 INCLUDES_PATH += $(BOLOS_SDK)/lib_cxng/include
 endif
 
-SOURCE_PATH   += $(BOLOS_SDK)/src $(foreach libdir, $(SDK_SOURCE_PATH), $(dir $(shell find $(BOLOS_SDK)/$(libdir) -name '*.[csS]'))) $(dir $(shell find $(APP_SOURCE_PATH) -name '*.[csS]'))
-SOURCE_FILES  += $(shell find $(SOURCE_PATH) -name '*.[csS]') $(GLYPH_DESTC) $(APP_SOURCE_FILES)
-INCLUDES_PATH += $(dir $(foreach libdir, $(SDK_SOURCE_PATH), $(dir $(shell find $(BOLOS_SDK)/$(libdir) -name '*.h')))) include $(BOLOS_SDK)/include $(BOLOS_SDK)/include/arm $(dir $(shell find $(APP_SOURCE_PATH) -name '*.h')) $(GLYPH_SRC_DIR)
+# Get absolute App root directory to ensure correct stem replacement
+APP_DIR := $(shell git rev-parse --show-toplevel)
+
+# Extract BOLOS_SDK source files added by the App Makefile
+APP_SRC_FROM_SDK = $(filter $(BOLOS_SDK)/%, $(APP_SOURCE_FILES))
+# Extract glyphs source files added by the App Makefile
+APP_SRC_GLYPH = $(filter $(GEN_SRC_DIR)/%, $(APP_SOURCE_FILES))
+# Extract proto source files added by the App Makefile
+APP_SRC_PROTOC = $(filter %.pb.c, $(APP_SOURCE_FILES) $(SOURCE_FILES)) $(shell find $(APP_SOURCE_PATH) -name '*.pb.c')
+# Filter remaining real source files from App (proto filtered globally hereafter)
+APP_SRC_FILTER = $(filter-out $(APP_SRC_FROM_SDK) $(APP_SRC_GLYPH), $(APP_SOURCE_FILES))
+
+# Separate SDK and APP, and GLYPHS sources
+SOURCES_SDK += $(foreach libdir, src $(SDK_SOURCE_PATH), $(shell find $(BOLOS_SDK)/$(libdir) -name '*.[csS]')) $(APP_SRC_FROM_SDK)
+SOURCES_APP += $(filter-out %.pb.c, $(abspath $(SOURCE_FILES) $(foreach libdir, $(SOURCE_PATH) $(APP_SOURCE_PATH), $(shell find $(libdir) -name '*.[csS]')) $(APP_SRC_FILTER)))
+SOURCES_GLYPHS += $(GLYPH_DESTC) $(APP_SRC_GLYPH)
+VPATH += $(call uniq, $(dir $(SOURCES_SDK) $(SOURCES_APP) $(GLYPH_DESTC)))
+
+# Retrieve APP header filenames
+INCLUDES_APP += $(shell find $(APP_SOURCE_PATH) -name '*.h')
+# Warn if a same header filename is found in both APP and SDK
+$(call check_duplicate, $(INCLUDES_APP), $(BOLOS_SDK))
```

Comment:
Can be interesting if we can warn for duplicates inside the app src itself

Reply:
I am not sure we should check the header files organization inside the Application itself. It should be something handled by the App owner directly, no :question:

Reply:
Ok, that not on us.